site/src/examples/carousel/Autoplay.tsx

```diff
@@ -0,0 +1,168 @@
+import {
+  Button,
+  FlexLayout,
+  Link,
+  StackLayout,
+  Text,
+  useId,
+} from "@salt-ds/core";
+import {
+  Carousel,
+  CarouselAnnouncement,
+  CarouselAutoplayIndicator,
+  CarouselCard,
+  type CarouselEmblaApiType,
+  CarouselNextButton,
+  CarouselPreviousButton,
+  CarouselProgressLabel,
+  CarouselSlides,
+} from "@salt-ds/embla-carousel";
+import { PauseIcon, PlayIcon } from "@salt-ds/icons";
+import type { EmblaCarouselType } from "embla-carousel";
+import { default as AutoplayPlugin } from "embla-carousel-autoplay";
+import Classnames from "embla-carousel-class-names";
+import { useCallback, useEffect, useRef, useState } from "react";
+import { sliderData } from "./exampleData";
+import styles from "./index.module.css";
+
+const DELAY_MSECS = 8000;
+const SETTLE_PIXEL_THRESHOLD = 10;
+
+export const Autoplay = () => {
+  const slideId = useId();
+  const [emblaApi, setEmblaApi] = useState<CarouselEmblaApiType | null>(null);
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [isPaused, setIsPaused] = useState(false);
+  const [slideIndex, setSlideIndex] = useState(0);
+
+  const autoplay = useRef(
+    AutoplayPlugin({ delay: DELAY_MSECS, playOnInit: false }),
+  );
+
+  const play = useCallback(() => {
+    autoplay.current.play();
+    setIsPlaying(true);
+    setIsPaused(false);
+  }, []);
+
+  const pause = useCallback(() => {
+    autoplay.current.stop();
+    setIsPaused(true);
+  }, []);
+
+  const stop = useCallback(() => {
+    autoplay.current.stop();
+    setIsPlaying(false);
+    setIsPaused(true);
+  }, []);
+
+  const handleSlideChange = useCallback((emblaApi: EmblaCarouselType) => {
+    const settledSlideIndex = emblaApi?.selectedScrollSnap() ?? 0;
+    setSlideIndex(settledSlideIndex + 1);
+  }, []);
+
+  useEffect(() => {
+    if (!emblaApi || !emblaApi.plugins()?.autoplay) {
+      return;
+    }
+    emblaApi.on("select", handleSlideChange);
+    return () => {
+      emblaApi.off("select", handleSlideChange);
+    };
+  }, [emblaApi, handleSlideChange]);
+
+  const timeUntilNext = autoplay.current.timeUntilNext() ?? DELAY_MSECS;
+
+  return (
+    <div className={"saltCarouselAutoplayExample"}>
+      <Carousel
+        aria-label="Autoplay example"
+        className={styles.carousel}
+        emblaOptions={{ loop: true, duration: 20 }}
+        emblaPlugins={[
+          autoplay.current,
```

Comment:
This is technically unsafe in concurrent React, is there a way to have it not read a ref during render?

Two ways come to mind:

Declare autoplay outside the component, not sure if this works?
Declare autoplay as state

Reply:
https://github.com/jpmorganchase/salt-ds/pull/5286#discussion_r2178631531